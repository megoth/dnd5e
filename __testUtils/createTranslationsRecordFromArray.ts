export default function createTranslationsRecordFromArray(
  url,
  list: string[]
): Record<string, string> {
  return list.reduce((memo: Record<string, string>, key) => {
    // eslint-disable-next-line no-param-reassign
    memo[`${url}#${key}`] = key;
    return memo;
  }, {});
}
