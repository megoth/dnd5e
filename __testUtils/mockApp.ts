import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
