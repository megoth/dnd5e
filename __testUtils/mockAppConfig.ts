export default function mockAppConfig() {
  return {
    errorsUrl: "http://localhost:3000/data/errors.ttl",
    translationsUrl: "http://localhost:3000/data/translations.ttl",
  };
}
