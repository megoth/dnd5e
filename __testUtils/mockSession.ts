import { Session } from "@inrupt/solid-client-authn-browser";

export interface SessionInfo {
  session: Session;
  sessionRequestInProgress: boolean;
  fetch: typeof window.fetch;
  login: () => Promise<void>;
  logout: () => Promise<void>;
}

export const authenticatedWebId = "http://example.com/webId#me";

export function mockAuthenticatedSession(options: Partial<Session> = {}) {
  return ({
    session: {
      info: {
        isLoggedIn: true,
        sessionId: "some-authenticated-session-id",
        webId: authenticatedWebId,
      },
      sessionRequestInProgress: false,
    },
    handleIncomingRedirect: () => Promise.resolve(),
    login: () => Promise.resolve(),
    logout: () => Promise.resolve(),
    on: () => {},
    ...options,
  } as unknown) as SessionInfo;
}

export function mockUnauthenticatedSession(options: Partial<Session> = {}) {
  return ({
    session: {
      info: {
        isLoggedIn: false,
        sessionId: "some-unauthenticated-session-id",
        webId: null,
      },
      sessionRequestInProgress: false,
    },
    handleIncomingRedirect: () => Promise.resolve(),
    login: options.login || jest.fn().mockResolvedValue(undefined),
    logout: () => Promise.resolve(),
    on: () => {},
    ...options,
  } as unknown) as SessionInfo;
}
