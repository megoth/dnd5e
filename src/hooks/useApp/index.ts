import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
