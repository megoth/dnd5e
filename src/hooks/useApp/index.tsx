import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
