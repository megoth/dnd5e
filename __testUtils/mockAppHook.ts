import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
