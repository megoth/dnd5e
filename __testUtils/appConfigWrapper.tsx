import React, { ReactNode } from "react";
import AppConfigProvider from "../src/contexts/appConfig";
import mockAppConfig from "./mockAppConfig";

interface Props {
  children: ReactNode;
}

export default function AppConfigWrapper({ children }: Props) {
  const { errorsUrl, translationsUrl } = mockAppConfig();
  return (
    <AppConfigProvider errorsUrl={errorsUrl} translationsUrl={translationsUrl}>
      {children}
    </AppConfigProvider>
  );
}
