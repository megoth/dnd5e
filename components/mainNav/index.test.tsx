import React from "react";
import { render } from "@testing-library/react";
import * as routerFns from "next/router";
import { NextRouter } from "next/router";
import solidUIReactFns from "@inrupt/solid-ui-react";
import { Session } from "@inrupt/solid-client-authn-browser";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import MainNav, { TESTID_MAIN_NAV_LIST_ITEM } from "./index";
import { mockAuthenticatedSession } from "../../__testUtils/mockSession";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("MainNav", () => {
  let mockedSessionHook;

  beforeEach(() => mockAppHook(mockedAppHook));
  beforeEach(() => {
    jest
      .spyOn(routerFns, "useRouter")
      .mockReturnValue({ asPath: "/" } as NextRouter);
  });
  beforeEach(() => {
    mockedSessionHook = jest.spyOn(solidUIReactFns, "useSession");
  });

  it("renders", () => {
    const { asFragment, getAllByTestId } = render(<MainNav />);
    expect(asFragment()).toMatchSnapshot();
    expect(getAllByTestId(TESTID_MAIN_NAV_LIST_ITEM)).toHaveLength(2);
  });

  it("renders an extra item for admin", () => {
    mockedSessionHook.mockReturnValue(
      mockAuthenticatedSession({
        session: {
          info: {
            isLoggedIn: true,
            sessionId: "some-authenticated-session-id",
            webId: "https://dnd5e.inrupt.net/profile/card#me", // hacky as hell for now
          },
        } as Session,
      })
    );
    const { asFragment, getAllByTestId } = render(<MainNav />);
    expect(asFragment()).toMatchSnapshot();
    expect(getAllByTestId(TESTID_MAIN_NAV_LIST_ITEM)).toHaveLength(3);
  });
});
