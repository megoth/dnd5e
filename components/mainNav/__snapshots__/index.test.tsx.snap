// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNav renders 1`] = `
<DocumentFragment>
  <nav
    class="pages-nav pages-nav--main-menu"
  >
    <ul
      class="pages-nav__list pages-nav__list--main-menu"
    >
      <li
        class="flex"
        data-testid="main-nav-list-item"
      >
        <a
          class="pages-nav__link pages-nav__link--main-menu"
          href="/characters"
        >
          [Translation for https://example.com/global/translations.ttl#charactersPageTitle does not exist]
        </a>
      </li>
      <li
        class="flex"
        data-testid="main-nav-list-item"
      >
        <a
          class="pages-nav__link pages-nav__link--main-menu"
          href="/rules"
        >
          [Translation for https://example.com/global/translations.ttl#rulesPageTitle does not exist]
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`MainNav renders an extra item for admin 1`] = `
<DocumentFragment>
  <nav
    class="pages-nav pages-nav--main-menu"
  >
    <ul
      class="pages-nav__list pages-nav__list--main-menu"
    >
      <li
        class="flex"
        data-testid="main-nav-list-item"
      >
        <a
          class="pages-nav__link pages-nav__link--main-menu"
          href="/characters"
        >
          [Translation for https://example.com/global/translations.ttl#charactersPageTitle does not exist]
        </a>
      </li>
      <li
        class="flex"
        data-testid="main-nav-list-item"
      >
        <a
          class="pages-nav__link pages-nav__link--main-menu"
          href="/rules"
        >
          [Translation for https://example.com/global/translations.ttl#rulesPageTitle does not exist]
        </a>
      </li>
      <li
        class="flex"
        data-testid="main-nav-list-item"
      >
        <a
          class="pages-nav__link pages-nav__link--main-menu"
          href="/admin"
        >
          [Translation for https://example.com/admin/translations.ttl#adminPageTitle does not exist]
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
